jdaviz/configs/mosviz/plugins/parsers.py

```diff
@@ -379,37 +380,84 @@ def mos_niriss_parser(app, data_dir, obs_label=""):
-    source_cat = sorted(list(p.glob("{}*_direct_*_cat.ecsv".format(obs_label))))
-    direct_image = sorted(list(p.glob("{}*_direct_dit1*_i2d.fits".format(obs_label))))
-    spec2d_r = sorted(list(p.glob("{}*_WFSSR_*_cal.fits".format(obs_label))))
-    spec2d_c = sorted(list(p.glob("{}*_WFSSC_*_cal.fits".format(obs_label))))
-    spec1d_r = sorted(list(p.glob("{}*_WFSSR_*_x1d.fits".format(obs_label))))
-    spec1d_c = sorted(list(p.glob("{}*_WFSSC_*_x1d.fits".format(obs_label))))
-
-    file_lists = {
-                  "Source Catalog": source_cat,
-                  "Direct Image": direct_image,
-                  "2D Spectra C": spec2d_c,
-                  "2D Spectra R": spec2d_r,
-                  "1D Spectra C": spec1d_c,
-                  "1D Spectra R": spec1d_r
-                 }
-
-    # Convert from pathlib Paths back to strings
-    for key in file_lists:
-        file_lists[key] = [str(x) for x in file_lists[key]]
+
+    file_types = ['Source Catalog', 'Direct Image', '2D Spectra C',
+                  '2D Spectra R', '1D Spectra C', '1D Spectra R']
+    file_lists = {k: [] for k in file_types}
+
+    # use FITS header keywords to sort the directory's files
+    for filepath in path.glob('*'):
+        if filepath.is_dir():
+            continue
+
+        filestr = str(filepath)
+
+        if filepath.suffix == '.fits':
+            # eligible files will have a DATAMODL value in their primary headers
+            header = fits.getheader(filepath, ext=0)
+            datamodl = header.get('DATAMODL')
+
+            # infer the dispersion direction of 1D and 2D spectra by the last
+            # letter of their filter values (e.g., "GR150C" or "GR150R")
+            try:
+                dispersion = header.get('FILTER')[-1]
+            except TypeError:
+                dispersion = None # could be cleaner...
+
+            # distinguish image files from counts files -- only the former go
+            # through the "Assign World Coordinate System" calibration step
+            s_wcs = header.get('S_WCS')
+
+            if datamodl is None:
+                continue
+            elif datamodl == 'MultiSpecModel' and dispersion == 'C':
+                file_lists['1D Spectra C'].append(filestr)
+            elif datamodl == 'MultiSpecModel' and dispersion == 'R':
+                file_lists['1D Spectra R'].append(filestr)
+            elif datamodl == 'MultiSlitModel' and dispersion == 'C':
+                file_lists['2D Spectra C'].append(filestr)
+            elif datamodl == 'MultiSlitModel' and dispersion == 'R':
+                file_lists['2D Spectra R'].append(filestr)
+            elif datamodl == 'ImageModel' and s_wcs is not None:
+                file_lists['Direct Image'].append(filestr)
+            else:
+                # print(f"Other case: datamodl is {datamodl}; "
+                #       f"dispersion is {dispersion}")
+                pass
```

Comment:
This case is here for my own testing purposes. I'll remove it entirely once the code is finalized since it would already be the last step of any iteration that reached it.